kanidmd/src/lib/idm/oauth2.rs

```diff
@@ -1018,4 +1136,141 @@ mod tests {
+                    audit,
+                    &ident,
+                    &uat,
+                    &consent_request.consent_token,
+                    ct,
+                )
+                .expect("Failed to perform oauth2 permit");
+
+            let client_authz = base64::encode(format!("test_resource_server:{}", secret));
+            let token_req = AccessTokenRequest {
+                grant_type: "authorization_code".to_string(),
+                code: permit_success.code,
+                redirect_uri: Url::parse("https://demo.example.com/oauth2/result").unwrap(),
+                client_id: None,
+                // From the first step.
+                code_verifier,
+            };
+
+            let token_response = idms_prox_read
+                .check_oauth2_token_exchange(audit, &client_authz, &token_req, ct)
+                .expect("Failed to perform oauth2 token exchange");
+
+            assert!(token_response.token_type == "bearer");
+
+            // Check a good req
+            let good_req = IntrospectionRequest {
+                token: token_response.access_token.clone(),
+                token_type_hint: None,
+            };
+            let r1 =
+                idms_prox_read.check_oauth2_token_introspect(audit, &client_authz, &good_req, ct);
+
+            let token = if let Ok(IntrospectionResponse::Active { active, token }) = r1 {
+                assert!(active);
+                token
+            } else {
+                panic!();
+            };
+
+            assert!(token.spn == uat.spn);
+
+            // Check some invalid requests
+            assert!(idms_prox_read
+                .check_oauth2_token_introspect(
+                    audit,
+                    &client_authz,
+                    &IntrospectionRequest {
+                        token: token_response.access_token.clone(),
+                        token_type_hint: Some("cheese".to_string()),
+                    },
+                    ct
+                )
+                .is_err());
+
+            assert!(idms_prox_read
+                .check_oauth2_token_introspect(
+                    audit,
+                    &client_authz,
+                    &IntrospectionRequest {
+                        token: "".to_string(),
+                        token_type_hint: None,
+                    },
+                    ct
+                )
+                .is_err());
+
+            // Now check if the time has passed to session exp
+            let exp = Duration::from_secs(token.oidc_token.exp as u64 + 1);
+
+            let r2 =
+                idms_prox_read.check_oauth2_token_introspect(audit, &client_authz, &good_req, exp);
+
+            if let Ok(IntrospectionResponse::Invalid { active }) = r2 {
+                assert!(!active);
+            } else {
+                panic!();
+            };
+
+            // expire the account, check it's enforced.
+            std::mem::drop(idms_prox_read);
```

Comment:
`drop` is in the prelude

```suggestion
            drop(idms_prox_read);
```